gmso/abc/abstract_site.py

```diff
@@ -0,0 +1,121 @@
+import warnings
+from typing import Union, Sequence, Any, TypeVar, ClassVar
+
+import numpy as np
+import unyt as u
+from unyt.exceptions import InvalidUnitOperation
+
+from pydantic import validator, root_validator, Field
+
+from gmso.abc.gmso_base import GMSOBase
+from gmso.exceptions import GMSOError
+
+PositionType = Union[Sequence[float], np.ndarray, u.unyt_array]
+SiteT = TypeVar('SiteT', bound='Site')
+
+BASE_DOC_ATTR = '__base_doc__'
+FIELDS_IN_DOCSTRING = 'alias_to_fields'
+
+
+class Site(GMSOBase):
+    __base_doc__: ClassVar[str] = """An interaction site object in the topology hierarchy.
+
+    Site is the object that represents any general interaction site in a molecular simulation.
+    Sites have been designed to be as general as possible, making no assumptions about representing atoms or beads, or
+    having mass or charge. That is, a Site can represent an atom in an atomistic system,
+    a bead in a coarse-grained system, and much more. 
+
+    Notes
+    -----
+    The label attribute for a site takes its meaning when used with some sort of container (like topology)
+    such that a label for a site can then be used to group sites together. The rules for defining a site label
+    and their meaning is left upto the container where the sites will reside. 
+    """
+
+    name_: str = Field(
+        '',
+        description='Name of the site, defaults to class name',
+    )
+
+    label_: str = Field(
+        '',
+        description='Label to be assigned for the site'
+    )
+
+    position_: PositionType = Field(
+        u.unyt_array([np.nan]*3, u.nm),
+        description='The Cartesian coordinates of the position of the site'
+    )
+
+    @property
+    def name(self) -> str:
+        return self.__dict__.get('name_')
+
+    @property
+    def position(self) -> PositionType:
+        return self.__dict__.get('position_')
+
+    @property
+    def label(self) -> str:
+        return self.__dict__.get('label_')
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__}, id {id(self)}>'
+
+    @validator('position_')
+    def is_valid_position(cls, position):
+        """Validator for attribute position"""
+        if position is None:
+            return u.unyt_array([np.nan] * 3, u.nm)
+
+        if not isinstance(position, u.unyt_array):
+            try:
+                position *= u.nm
+            except InvalidUnitOperation as e:
+                raise GMSOError('Converting object of type {0} failed with following error: {1}'
```

Comment:
We use an `fstring` in the ValueError below, could we use fstrings here, or is there a reason not to?

Reply:
No, not really, this was copied as is from the code in master,